peps/pep-0728.rst

```diff
@@ -0,0 +1,614 @@
+
+``__extra__`` specifies the value type for fields of keys that are *unknown* to
+the TypedDict type. So the value type of any known field is not necessarily
+consistent with ``__extra__``'s type, and ``__extra__``'s type is not
+necessarily consistent with the value types of all known fields.
+
+This differs from TypeScript's Index Signatures [6]_ syntax, which requires all
+properties' types to match the string index's type. For example:
+
+.. code-block:: typescript
+
+    interface MovieWithExtraNumber {
+        name: string // Property 'name' of type 'string' is not assignable to 'string' index type 'number'.
+        [index: string]: number
+    }
+
+    interface MovieWithExtraNumberOrString {
+        name: string // OK
+        [index: string]: number | string
+    }
+
+This is a known limitation is discussed in TypeScript's issue tracker [7]_,
+where it is suggested that there should be a way to exclude the defined keys
+from the index signature, so that it is possible to define a type like
+``MovieWithExtraNumber``.
+
+A workaround is to use intersection:
+
+.. code-block:: typescript
+
+    interface MovieExtras {
+        [index: string]: number
+    }
+
+    interface MovieBase {
+        name: string
+    }
+
+    type Movie = MovieExtras & MovieBase
+
+However, this does not work for assignment:
+
+.. code-block:: typescript
+
+    const movie: Movie = {
+        name: "No Country for Old Men", // Type 'string' is not assignable to type 'number'
+        year: 2007,
+    }
+
+How Else Will This Interact with PEP 692?
+-----------------------------------------
+
+Quoting from a comment on a Pull Request that implements part of :pep:`692` [8]_
+in mypy, we want it to be as simple as possible, reusing the existing type
+checking logic:
+
+    There is nothing cumbersome in reducing the PEP to just one paragraph that
+    would explain that `Unpack[SomeTD]` is a syntactic sugar for (and is
+    considered equivalent to) the expanded signature. This has a number of
+    benefits:
+
+    * This will not add any new unsafety that is not already present for
+      existing uses of TypedDicts in `**` contexts. (And type checkers may
+      handle this unsafety in a uniform way, say in mypy we _may_ use existing
+      `--extra-checks` flag to prohibit some techincally unsafe calls as I
+      mentioned before.)
+
+    * This is actually easy to remember and to reason about.
+
+    * This will allow people who want subtyping between callables to easily
+      achieve this using `total=False`, which follows from existing rules for
+      expanded callables.
+
+Backwards Compatibility
+=======================
+
+Dunder attributes like ``__extra__`` are reserved for stdlib, we need not
+concerned about using it in this proposal. If the proposal is accepted, neither
+``__required_keys__`` and ``__optional_keys__`` should include ``__extra__`` in
+runtime.
+
+Because this is a type-checking feature, it can be made available to older
```

Comment:
Or even without typing_extensions. The PEP as currently written would require few if any changes to the runtime implementation.

Reply:
Also regarding alternatives to `__extra__`, I think something like this will not require changes to the runtime implementation at all:
```
    class Foo(TypedDict):
        def __extra_field__(self, key: str) -> int:
            ...
```